ConstExpr: Significant progress to supporting more complicated cases

lattner opened the pull request:
- Add a specific diagnostic for loops, which we don't support yet (intentionally).
 - Implement support for copy_addr, begin_access, end_access, destroy_addr.
 - Add support for ptrtoint of strings to word size, and llvm.expect.
 - Fix updateIndexedElement to be (a) correct, and (b) handle overwriting
   subelements of uninit memory, which happens with piecewise initialization
   of constant values.

Many of these cases are exercised by SignedInteger<>.init<A>(_:) which is
incredibly complicated.

Unfortunately, this isn't enough to make either testcase work, because I still
have a bug dealing with some witness table applications.  This isn't a
regression, but will need further work before those tests light up.

lattner committed:
ConstExpr: Significant progress to supporting more complicated cases:
 - Add a specific diagnostic for loops, which we don't support yet (intentionally).
 - Implement support for copy_addr, begin_access, end_access, destroy_addr.
 - Add support for ptrtoint of strings to word size, and llvm.expect.
 - Fix updateIndexedElement to be (a) correct, and (b) handle overwriting
   subelements of uninit memory, which happens with piecewise initialization
   of constant values.

Many of these cases are exercised by SignedInteger<>.init<A>(_:) which is
incredibly complicated.

Unfortunately, this isn't enough to make either testcase work, because I still
have a bug dealing with some witness table applications.  This isn't a
regression, but will need further work before those tests light up.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow, srsly

lattner commented:
@swift-ci please test tensorflow

lattner committed:
Fixes to verify things are the correct type before just using the value.
With our new representation we can end up with strings having integer
type.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -824,29 +930,56 @@ ConstExprFunctionCache::evaluateFlowSensitive(SILInstruction *inst) {
+  }
+
+  // Copy addr is a load + store combination.
+  if (auto *copy = dyn_cast<CopyAddrInst>(inst)) {
```
ACAICT, the number of `dyn_cast` cases is growing within `ConstExprFunctionCache::evaluateFlowSensitive`. Although a visitor is too heavyweight for this, is it worth turning them into a switch stmt? You didn't use methods specific to subclasses, so a switch seems to work better.

lattner replied:
This is the standard pattern.  Are you concerned about compile time overheads due to repeated if/else chains?   Years ago I did work to make sure the LLVM optimizer turned this into a switch on opcode.  It is possible that someone broke it though.

rxwei replied:
K! That's good to know.

mhong commented on test/SILOptimizer/pound_assert.swift:
```diff
@@ -19,15 +26,41 @@ func nonConstant() {
+  // expected-error @+1 {{#assert condition not constant}}
+  #assert(loops1(a: 20000) > 42)
+
+  // TODO: xpected-error @+1 {{#assert condition not constant}}
```
typo

lattner replied:
This is intentional, the test is disabled (that's the TODO).

lattner committed:
Fix a crasher in lookupOrLinkWitnessTable which occurred when it got
an abstract conformance.  This was the thing breaking a bot.

Add support for DebugValueAddrInst, improve logging, add another
(failing) testcase reduced from the stdlib.

lattner commented:
@swift-ci please test tensorflow

lattner committed:
A lot of progress handling substitution mappings from callers to callees.
This is still hacky, and I have questions out on swift-dev.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
My git-fu is too weak to figure out how to rebase this onto latest code in the tensorflow branch, necessary to unfail those tests.  Moving over to PR17298.
